contrib/config/source/kv_store_xds_delegate.cc

```diff
@@ -0,0 +1,127 @@
+#include "contrib/config/source/kv_store_xds_delegate.h"
+
+#include "envoy/registry/registry.h"
+#include "envoy/service/discovery/v3/discovery.pb.h"
+
+#include "source/common/common/logger.h"
+#include "source/common/config/utility.h"
+#include "source/common/protobuf/utility.h"
+
+#include "absl/strings/match.h"
+#include "absl/strings/str_cat.h"
+#include "absl/types/optional.h"
+#include "contrib/envoy/extensions/config/v3alpha/kv_store_xds_delegate_config.pb.h"
+#include "contrib/envoy/extensions/config/v3alpha/kv_store_xds_delegate_config.pb.validate.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Config {
+namespace {
+
+using ::Envoy::Config::XdsSourceId;
+using ::envoy::extensions::config::v3alpha::KeyValueStoreXdsDelegateConfig;
+
+// TODO(abeyad): Add a per-Envoy instance prefix to the storage key, to allow xDS resources to be
+// persisted per-running-Envoy when Envoy is embedded as a library (e.g. Envoy Mobile).
+
+// Constructs the key for the resource, to be used in the KeyValueStore.
+std::string constructKey(const XdsSourceId& source_id, const std::string& resource_name) {
+  static constexpr char DELIMITER[] = "+";
+  return absl::StrCat(source_id.toKey(), DELIMITER, resource_name);
+}
+
+} // namespace
+
+KeyValueStoreXdsDelegate::KeyValueStoreXdsDelegate(KeyValueStorePtr&& xds_config_store)
+    : xds_config_store_(std::move(xds_config_store)) {}
+
+std::vector<envoy::service::discovery::v3::Resource>
+KeyValueStoreXdsDelegate::getResources(const XdsSourceId& source_id,
+                                       const std::vector<std::string>& resource_names) const {
+  if (resource_names.empty() || (resource_names.size() == 1 && resource_names[0] == "*")) {
+    // Empty names or one entry with "*" means wildcard.
```

Comment:
optional nit: I mildly dislike having two wildcard options.  Could we validate that the resource name not be empty?

Reply:
Unfortunately, Envoy allows these two different ways to specify wildcard (empty resource names or one resource name with "*").  Or do you mean, when invoking getResources, if resource_names is empty, add one `*` resource name to the list before passing in?

Reply:
no I didn't know this was prior art.  Don't mind me then :-P